### ISSUE TITLE
Trailing whitespace is included for endloc calculation even though it is not matched.

### ISSUE THREAD

**Issue by kdop** [pyparsing/pyparsing#344]
Hi,

I am using Python 3.9.5 and pyparsing==3.0.6

I have a scenario where I need to match a word(alphanum) and optionally another word. However, pp.Optional() combined with trailing whitespace leads to incorrect endloc index:

```
import pyparsing as pp


def first_match(expression, text):
    for m, s, e in expression.scanString(text, maxMatches=1):
        return {'match': m, 'start': s, 'end': e}
    return None


expr = pp.Group(
    pp.Word(pp.alphanums) +
    pp.Optional(
        pp.Word(pp.alphanums)
    )
)
```

Use:

1. No trailing whitespace - everything is as expected:

```
print(first_match(expr, "one"))
# {'match': ParseResults([ParseResults(['one'], {})], {}), 'start': 0, 'end': 3}
```

2. Trailing white space - notice matched token (still 'one') and end location which is now 8.

```
print(first_match(expr, "one     "))
# {'match': ParseResults([ParseResults(['one'], {})], {}), 'start': 0, 'end': 8}
```

Shouldn't the endloc index returned by scanString always be the index of the last match character?

I have opened same issue in stackoverflow: https://stackoverflow.com/questions/70291338/pyparsing-trailing-whitespace-is-not-matched-but-still-included-for-endloc-calc

**Comment by ptmcg** [comment]
Please don't duplicate issues in Stack Overflow.

**Comment by ptmcg** [comment]
This is answered on SO: https://stackoverflow.com/questions/70291338/pyparsing-trailing-whitespace-is-not-matched-but-still-included-for-endloc-calc/70416366#70416366
